docs/source/en/llm_optims.md

```diff
@@ -0,0 +1,324 @@
+['The theory of special relativity states 1. The speed of light is constant in all inertial reference']
+```
+
+</hfoption>
+<hfoption id="setup_cache">
+
+> [!WARNING]
+> The `_setup_cache` method is an internal method and still under development, so keep in mind this will likely change in the future.
+
+The `_setup_cache` method doesn't support [`GenerationMixin.generate`] yet, so this method is a bit more involved. You'll need to write your own function to decode the next token given the current token and position and cache position of previously generated tokens.
+
+```py
+from transformers import LlamaTokenizer, LlamaForCausalLM, StaticCache, logging
+from transformers.testing_utils import CaptureLogger
+import torch
+
+prompts = [
+    "Simply put, the theory of relativity states that ",
+    "My favorite all time favorite condiment is ketchup.",
+]
+
+NUM_TOKENS_TO_GENERATE = 40
+torch_device = "cuda"
+
+tokenizer = LlamaTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf", pad_token="</s>", padding_side="right")
+model = LlamaForCausalLM.from_pretrained("meta-llama/Llama-2-7b-hf", device_map="sequential")
+inputs = tokenizer(prompts, return_tensors="pt", padding=True).to(model.device)
+
+def decode_one_tokens(model, cur_token, input_pos, cache_position):
+    logits = model(
+        cur_token, position_ids=input_pos, cache_position=cache_position, return_dict=False, use_cache=True
+    )[0]
+    new_token = torch.argmax(logits[:, -1], dim=-1)[:, None]
+    return new_token
+```
+
+The are two important things you must do to enable static kv-cache and torch.compile with the `_setup_cache` method:
+
+1. Access the model's `_setup_cache` method and pass it the [`StaticCache`] class. This is a more flexible method because it allows you to configure parameters like the maximum batch size and sequence length.
+
+2. Call torch.compile on the model to compile the forward pass with the static kv-cache.
+
+```py
+batch_size, seq_length = inputs["input_ids"].shape
+with torch.no_grad():
+     model._setup_cache(StaticCache, 2, max_cache_len=4096)
+     cache_position = torch.arange(seq_length, device=torch_device)
+     generated_ids = torch.zeros(
+         batch_size, seq_length + NUM_TOKENS_TO_GENERATE + 1, dtype=torch.int, device=torch_device
+     )
+     generated_ids[:, cache_position] = inputs["input_ids"].to(torch_device).to(torch.int)
+
+     logits = model(**inputs, cache_position=cache_position, return_dict=False, use_cache=True)[0]
+     next_token = torch.argmax(logits[:, -1], dim=-1)[:, None]
+     generated_ids[:, seq_length] = next_token[:, 0]
+
+     decode_one_tokens = torch.compile(decode_one_tokens, mode="reduce-overhead", fullgraph=True)
+     cache_position = torch.tensor([seq_length + 1], device=torch_device)
+     for _ in range(1, NUM_TOKENS_TO_GENERATE):
+         with torch.backends.cuda.sdp_kernel(enable_flash=False, enable_mem_efficient=False, enable_math=True):
```

Comment:
It would be good to have a comment here explaining the value of this context manager